hnn_core/parallel_backends.py

```diff
@@ -588,26 +610,23 @@ def __init__(self, n_procs=None, mpi_cmd='mpiexec'):
 
         hyperthreading = False
 
-        try:
-            import mpi4py
-            mpi4py.__version__  # for flake8 test
+        if _has_mpi4py() and _has_psutil():
+            import psutil
 
-            try:
-                import psutil
+            n_physical_cores = psutil.cpu_count(logical=False)
 
-                n_physical_cores = psutil.cpu_count(logical=False)
-
-                # detect if we need to use hwthread-cpus with mpiexec
-                if self.n_procs > n_physical_cores:
-                    hyperthreading = True
-
-            except ImportError:
-                warn('psutil not installed, so cannot detect if hyperthreading'
-                     'is enabled, assuming yes.')
+            # detect if we need to use hwthread-cpus with mpiexec
+            if self.n_procs > n_physical_cores:
                 hyperthreading = True
 
-        except ImportError:
-            warn('mpi4py not installed. will run on single processor')
+        else:
+            packages = list()
+            if not _has_mpi4py():
+                packages += ['mpi4py']
+            if not _has_psutil():
+                packages += ['psutil']
```

Comment:
Why do `_has_mpi4py()` and `_has_psutil()` need to be run twice? Couldn't we just use nested 'if' statements?

Reply:
In the old code, `psutil` import statement was nested because it was supposed to be optional but it's used in the other functions for killing the `nrniv` process. I made a judgement call to make `psutil` a requirement if you are using MPIBackend. 

Now, we need the second `_has_mpi4py` only to print the warning message. We could assign it to a variable to avoid importing twice but probably not worth the effort

Reply:
I think what I was originally imagining won't work. This looks good!